contracts/src/v0.8/shared/enumerable/CCIPEnumerableMap.sol

```diff
@@ -0,0 +1,143 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {EnumerableSet} from "../../vendor/openzeppelin-solidity/v4.8.3/contracts/utils/structs/EnumerableSet.sol";
+
+/**
+ * @dev Library for managing an enumerable variant of Solidity's
+ * https://solidity.readthedocs.io/en/latest/types.html#mapping-types[`mapping`]
+ * type.
+ *
+ * Maps have the following properties:
+ *
+ * - Entries are added, removed, and checked for existence in constant time
+ * (O(1)).
+ * - Entries are enumerated in O(n). No guarantees are made on the ordering.
+ *
+ * ```
+ * contract Example {
+ *     // Add the library methods
+ *     using CCIPEnumerableMap for CCIPEnumerableMap.Bytes32ToBytesMap;
+ *
+ *     // Declare a set state variable
+ *     CCIPEnumerableMap.Bytes32ToBytesMap private myMap;
+ * }
+ * ```
+ *
+ * The following map types are supported:
+ *
+ * - `bytes32 -> bytes` (`Bytes32ToBytes`)
+ *
+ * [WARNING]
+ * ====
+ * Trying to delete such a structure from storage will likely result in data corruption, rendering the structure
+ * unusable.
+ * See https://github.com/ethereum/solidity/pull/11843[ethereum/solidity#11843] for more info.
+ *
+ * In order to clean an CCIPEnumerableMap, you can either remove all elements one by one or create a fresh instance using an
+ * array of CCIPEnumerableMap.
+ * ====
+ */
+library CCIPEnumerableMap {
+  using EnumerableSet for EnumerableSet.Bytes32Set;
+
+  // Define a custom error
+  error NonexistentKeyError();
+
+  // To implement this library for multiple types with as little code
+  // repetition as possible, we write it in terms of a generic Map type with
+  // bytes32 keys and values.
+  // The Map implementation uses private functions, and user-facing
+  // implementations (such as Bytes32ToBytesMap) are just wrappers around
+  // the underlying Map.
+  struct Bytes32ToBytesMap {
+    EnumerableSet.Bytes32Set _keys;
+    mapping(bytes32 => bytes) _values;
+  }
+
+  /**
+   * @dev Adds a key-value pair to a map, or updates the value for an existing
+   * key. O(1).
+   *
+   * Returns true if the key was added to the map, that is if it was not
+   * already present.
+   */
+  // solhint-disable-next-line chainlink-solidity/prefix-internal-functions-with-underscore
+  function set(Bytes32ToBytesMap storage map, bytes32 key, bytes memory value) internal returns (bool) {
+    map._values[key] = value;
+    return map._keys.add(key);
+  }
+
+  /**
+   * @dev Removes a key-value pair from a map. O(1).
+   *
+   * Returns true if the key was removed from the map, that is if it was present.
+   */
+  // solhint-disable-next-line chainlink-solidity/prefix-internal-functions-with-underscore
+  function remove(Bytes32ToBytesMap storage map, bytes32 key) internal returns (bool) {
+    delete map._values[key];
+    return map._keys.remove(key);
```

Comment:
q: does this correctly remove all bytes values, since `bytes` is a dynamic array?

Reply:
added a test for this. will test on varying bytes size